File: services/api/app/auth/iap.py
from dataclasses import dataclass

from google.auth.transport import requests
from google.oauth2 import id_token

from app.config import settings

IAP_HEADER = "x-goog-iap-jwt-assertion"


@dataclass(frozen=True)
class IapUser:
    email: str
    sub: str


class IapAuthError(Exception):
    pass


def _email_domain(email: str) -> str:
    _, _, domain = email.rpartition("@")
    return domain.lower()


def validate_iap_jwt(token: str) -> IapUser:
    if settings.iap_jwt_validation_disabled:
        if settings.environment == "prod":
            raise IapAuthError("IAP JWT validation is disabled in prod")
        return IapUser(email="local-dev@internal", sub="local-dev")

    if not settings.iap_audience:
        raise IapAuthError("IAP audience not configured")

    try:
        info = id_token.verify_oauth2_token(
            token, requests.Request(), audience=settings.iap_audience
        )
    except ValueError as exc:
        raise IapAuthError("Invalid IAP JWT") from exc

    email = info.get("email")
    sub = info.get("sub")
    if not email or not sub:
        raise IapAuthError("IAP JWT missing identity claims")

    if not info.get("email_verified", False):
        raise IapAuthError("IAP JWT email is not verified")

    allowed_domains = settings.parsed_allowed_email_domains()
    if allowed_domains:
        domain = _email_domain(email)
        hosted_domain = info.get("hd")
        if domain not in allowed_domains and hosted_domain not in allowed_domains:
            raise IapAuthError("Email domain is not allowed")

    return IapUser(email=email.lower(), sub=sub)

File: services/api/internal/auth/iap.go
package auth

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"google.golang.org/api/idtoken"

	"github.com/therealagt/ContractManagementTool/services/api/internal/config"
)

const IAPHeader = "x-goog-iap-jwt-assertion"

var ErrIAPAuth = errors.New("iap auth error")

type IAPUser struct {
	Email string
	Sub   string
}

type IAPValidator struct {
	settings *config.Settings
}

func NewIAPValidator(settings *config.Settings) *IAPValidator {
	return &IAPValidator{settings: settings}
}

func (v *IAPValidator) devBypassUser() *IAPUser {
	if v.settings.IAPJWTValidationDisabled && v.settings.Environment != "prod" {
		return &IAPUser{Email: "local-dev@internal", Sub: "local-dev"}
	}
	return nil
}

func (v *IAPValidator) Validate(ctx context.Context, token string) (*IAPUser, error) {
	if user := v.devBypassUser(); user != nil {
		return user, nil
	}

	if v.settings.IAPAudience == "" {
		return nil, fmt.Errorf("%w: IAP audience not configured", ErrIAPAuth)
	}

	payload, err := idtoken.Validate(ctx, token, v.settings.IAPAudience)
	if err != nil {
		return nil, fmt.Errorf("%w: invalid IAP JWT", ErrIAPAuth)
	}

	email, _ := payload.Claims["email"].(string)
	sub, _ := payload.Claims["sub"].(string)
	if email == "" || sub == "" {
		return nil, fmt.Errorf("%w: IAP JWT missing identity claims", ErrIAPAuth)
	}

	verified, _ := payload.Claims["email_verified"].(bool)
	if !verified {
		return nil, fmt.Errorf("%w: IAP JWT email is not verified", ErrIAPAuth)
	}

	allowed := v.settings.ParsedAllowedEmailDomains()
	if len(allowed) > 0 {
		domain := emailDomain(email)
		hostedDomain, _ := payload.Claims["hd"].(string)
		if !contains(allowed, domain) && !contains(allowed, strings.ToLower(hostedDomain)) {
			return nil, fmt.Errorf("%w: email domain is not allowed", ErrIAPAuth)
		}
	}

	return &IAPUser{Email: strings.ToLower(email), Sub: sub}, nil
}

func emailDomain(email string) string {
	_, domain, ok := strings.Cut(email, "@")
	if !ok {
		return ""
	}
	return strings.ToLower(domain)
}

func contains(items []string, target string) bool {
	for _, item := range items {
		if item == target {
			return true
		}
	}
	return false
}
